src/wallet/mod.rs

```diff
@@ -387,60 +394,31 @@ where
         Ok(new_addresses_cached)
     }
 
-    /// Return vector of n unused addresses from the [`KeychainKind`].
-    /// If less than n unused addresses are returned, the rest will be populated by new addresses.
-    /// The unused addresses returned are in order of oldest in keychain first, with increasing index.
-    pub fn get_batch_unused_addresses(
-        &self,
-        n: usize,
-        from_front: bool,
-        keychain: KeychainKind,
-    ) -> Result<Vec<AddressInfo>, Error> {
+    /// Return set of unused script indexes for the [`KeychainKind`].
+    pub fn get_unused_script_indexes(&self, keychain: KeychainKind) -> Result<Vec<u32>, Error> {
         let script_pubkeys = self
             .database
             .borrow()
             .iter_script_pubkeys(Some(keychain))
             .unwrap_or_else(|_| vec![]);
-
         let current_address_index = self.fetch_index(keychain)? as usize;
-        let check_indexes = if from_front {
-            (0..=current_address_index).collect::<Vec<_>>()
-        } else {
-            (0..=current_address_index).rev().collect::<Vec<_>>()
-        };
+        let txns = self.list_transactions(true).unwrap_or_else(|_| vec![]);
 
-        let mut unused_addresses = vec![];
-        for i in check_indexes {
-            // if we have made a pubkey at this index, check whether the address has been used.
-            if i < script_pubkeys.len() {
-                let script_pk = &script_pubkeys[i];
-                if self.has_address_been_used(script_pk) {
-                    continue;
-                }
-            }
-            if let Ok(unused_address) = self
-                .get_descriptor_for_keychain(keychain)
-                .as_derived(i as u32, &self.secp)
-                .address(self.network)
-                .map(|address| AddressInfo {
-                    address,
-                    index: i as u32,
-                    keychain,
-                })
-                .map_err(|_| Error::ScriptDoesntHaveAddressForm)
-            {
-                unused_addresses.push(unused_address);
-            }
+        let txn_scripts: HashSet<Script> = txns
+            .iter()
+            .flat_map(|tx_details| tx_details.transaction.as_ref())
+            .flat_map(|tx| tx.output.iter())
+            .map(|o| o.script_pubkey.clone())
+            .collect();
 
-            if unused_addresses.len() >= n {
-                break;
+        let mut scripts_not_used = vec![];
+        for i in 0..=current_address_index {
+            if (i < script_pubkeys.len()) && txn_scripts.contains(&script_pubkeys[i]) {
+            } else {
+                scripts_not_used.push(i as u32);
             }
         }
-
-        for _ in 0..(n - unused_addresses.len()) {
-            unused_addresses.push(self.get_new_address(keychain)?)
-        }
-        Ok(unused_addresses)
+        Ok(scripts_not_used.into_iter().collect())
```

Comment:
Looks like this `into_iter` and `collect` not necessary here since you just pushed them into `vec`.